vllm/model_executor/models/qwen2_audio.py

```diff
@@ -0,0 +1,453 @@
+    output_lengths = (max_source_position - 2) // 2 + 1
+    return output_lengths
+
+
+def input_processor_for_qwen2_audio(
+        ctx: InputContext, inputs: DecoderOnlyInputs) -> DecoderOnlyInputs:
+    multi_modal_data = inputs.get("multi_modal_data")
+    if multi_modal_data is None or "audio" not in multi_modal_data:
+        return inputs
+    if len(multi_modal_data["audio"]) == 0:
+        return inputs
+    assert (isinstance(multi_modal_data['audio'], list)
+            and isinstance(multi_modal_data['audio'][0], tuple))
+
+    processor = cached_get_processor(ctx.model_config.model)
+    audios = [_[0] for _ in multi_modal_data['audio']]
+
+    audio_inputs = processor.feature_extractor(audios,
+                                               sampling_rate=16000,
+                                               return_attention_mask=True,
+                                               padding="max_length")
+
+    feature_attention_mask = audio_inputs.pop("attention_mask")
+    input_features = audio_inputs['input_features']
+
+    audio_feat_lengths, audio_output_lengths = _get_feat_extract_output_lengths(
+        feature_attention_mask.sum(-1))
+
+    audio_token_index = ctx.model_config.hf_config.audio_token_index
+
+    input_ids = inputs['prompt_token_ids']
+
+    new_input_ids = []
+    audio_num = input_ids.count(audio_token_index)
+    assert len(input_features) == audio_num, \
+        (f'The text input contains {audio_num} audio tokens, '
+         f'but {len(input_features)} audios provided')
+    start = 0
+    for audio_idx in range(audio_num):
+        end = input_ids.index(audio_token_index, start)
+        new_input_ids.extend(input_ids[start:end])  # text part
+
+        new_input_ids.extend([audio_token_index] *
+                             audio_output_lengths[audio_idx])
+        start = end + 1
+    new_input_ids.extend(input_ids[start:])
+
+    return token_inputs(
+        prompt_token_ids=new_input_ids,
+        prompt=inputs['prompt'],
+        multi_modal_data=multi_modal_data,
+    )
+
+
+def input_mapper_for_qwen2_audio(
+    ctx: InputContext,
+    multi_modal_data: Union[np.ndarray, List[np.ndarray]],
+) -> MultiModalInputs:
+    """Input mapper for Qwen2-Audio."""
+    processor = cached_get_processor(ctx.model_config.model)
+    audio_feature_extractor = processor.feature_extractor
+    if audio_feature_extractor is None:
+        raise RuntimeError(
+            "No HuggingFace audio_feature_extractor is available "
+            "to process the audio object")
+
+    if len(multi_modal_data) == 0:
+        batch_data = {
+            'input_features': torch.tensor([]),
+            'feature_attention_mask': torch.tensor([])
+        }
+        return batch_data
+    try:
+        audios = [_[0] for _ in multi_modal_data]
+        batch_data = audio_feature_extractor(audios,
+                                             sampling_rate=16000,
```

Comment:
Is it intended that you don't use the sampling rate in `multi_modal_data` (i.e. `_[1]` in the list comprehension) when calling the `audio_feature_extractor`?

Reply:
Yes, we don't use the sampling rate, because the sampling rate is always 16000.